### PR TITLE
AMD ROCm 4.2.0: Bump up rocm recipes part 2

### PR DESCRIPTION
**Pull request by arjun-raj-kuppala**
AMD ROCm 4.2.0: Bump up rocm-cmake hsakmt-roct rocm-smi-lib hsa-rocr-dev llvm-amdgpu rocm-debug-agent rocm-gdb rocalution rocm-validation-suite

### CODE REVIEW

**Review by haampie — CHANGES_REQUESTED** [review]
1. Can you remove all occurrences of if's in for loops and just turn them into two for loops? It is better to make package.py's declarative and reduce control flow to a minimum.

2. I'm rather confused by the fact that clang is now aware of spack and how that went through reviews in llvm; skimming through the code it seems it makes multiple incorrect assumptions about the folder structure. I don't think it's a good idea at all for clang to be aware of spack, but well... it's released already.

File: var/spack/repos/builtin/packages/hsa-rocr-dev/package.py [hunk 1]
```diff
@@ -49,7 +53,18 @@ def cmake_args(self):
         libelf_include = self.spec['libelf'].prefix.include.libelf
         args = ['-DLIBELF_INCLUDE_DIRS=%s' % libelf_include]
 
-        if '@3.7.0:' in self.spec:
+        if '+shared' in self.spec:
+            args.append('-DBUILD_SHARED_LIBS=ON')
+        else:
+            args.append('-DBUILD_SHARED_LIBS=OFF')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/hsa-rocr-dev/package.py]: Can you use `self.define_from_variant(.., ..)` where applicable to reduce code bloat?

**Review by haampie — CHANGES_REQUESTED** [review]

**Comment by arjun-raj-kuppala** [comment]
> 1. Can you remove all occurrences of if's in for loops and just turn them into two for loops? It is better to make package.py's declarative and reduce control flow to a minimum.**(I have addressed this with new patch)**
> 2. I'm rather confused by the fact that clang is now aware of spack and how that went through reviews in llvm; skimming through the code it seems it makes multiple incorrect assumptions about the folder structure. I don't think it's a good idea at all for clang to be aware of spack, but well... it's released already.**(I will pass this suggestion to the concerned team.)**

**Review by haampie — DISMISSED** [review]

File: var/spack/repos/builtin/packages/hsa-rocr-dev/package.py [hunk 2]
```diff
@@ -49,7 +53,13 @@ def cmake_args(self):
...
 
+        self.define_from_variant('BUILD_SHARED_LIBS', 'shared')
+
         if '@3.7.0:' in self.spec:
-            args.append('-DIMAGE_SUPPORT=ON')
+            self.define_from_variant('IMAGE_SUPPORT', 'image')
+
+        if '@4.2.0:' in self.spec:
+            bitcode_dir = self.spec['rocm-device-libs'].prefix.amdgcn.bitcode
+            args.append('-DBITCODE_DIR={}'.format(bitcode_dir))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/hsa-rocr-dev/package.py]:
```suggestion
            args.append('-DBITCODE_DIR={0}'.format(bitcode_dir))
```

**Review by haampie — APPROVED** [review]

**Pull request merged**